File: backend/app/services/report.py
from __future__ import annotations

from datetime import datetime
from io import BytesIO
from xml.sax.saxutils import escape

from reportlab.lib import colors
from reportlab.lib.enums import TA_LEFT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import inch
from reportlab.platypus import Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle


def _text(value) -> str:
    if value is None:
        return "-"
    if isinstance(value, (list, tuple)):
        return ", ".join(str(item) for item in value if item) or "-"
    return str(value).strip() or "-"


def _paragraph(text: str, style: ParagraphStyle) -> Paragraph:
    return Paragraph(escape(_text(text)).replace("\n", "<br/>"), style)


def _labeled_paragraph(label: str, value, style: ParagraphStyle) -> Paragraph:
    safe_value = escape(_text(value)).replace("\n", "<br/>")
    return Paragraph(f"<b>{escape(label)}:</b> {safe_value}", style)


def _format_datetime(value: datetime | None) -> str:
    if not value:
        return "-"
    return value.strftime("%Y-%m-%d %H:%M:%S")


def _average_score(turns: list) -> str:
    scores = []
    for turn in turns:
        evaluation = turn.evaluation or {}
        score = evaluation.get("score")
        if isinstance(score, (int, float)):
            scores.append(score)
    if not scores:
        return "-"
    return f"{sum(scores) / len(scores):.1f} / 10"


def build_session_report_pdf(session, max_turns: int) -> bytes:
    buffer = BytesIO()
    doc = SimpleDocTemplate(
        buffer,
        pagesize=A4,
        leftMargin=0.65 * inch,
        rightMargin=0.65 * inch,
        topMargin=0.6 * inch,
        bottomMargin=0.6 * inch,
    )

    styles = getSampleStyleSheet()
    title_style = ParagraphStyle(
        "ReportTitle",
        parent=styles["Title"],
        fontName="Helvetica-Bold",
        fontSize=22,
        leading=26,
        textColor=colors.HexColor("#172033"),
        alignment=TA_LEFT,
        spaceAfter=12,
    )
    section_style = ParagraphStyle(
        "SectionTitle",
        parent=styles["Heading2"],
        fontName="Helvetica-Bold",
        fontSize=14,
        leading=18,
        textColor=colors.HexColor("#1d4ed8"),
        spaceBefore=10,
        spaceAfter=8,
    )
    body_style = ParagraphStyle(
        "Body",
        parent=styles["BodyText"],
        fontName="Helvetica",
        fontSize=10.5,
        leading=15,
        textColor=colors.HexColor("#172033"),
        spaceAfter=6,
    )
    label_style = ParagraphStyle(
        "Label",
        parent=body_style,
        fontName="Helvetica-Bold",
        textColor=colors.HexColor("#334155"),
    )
    small_style = ParagraphStyle(
        "Small",
        parent=body_style,
        fontSize=9,
        leading=12,
        textColor=colors.HexColor("#526071"),
    )

    completed_turns = sorted([turn for turn in session.turns if turn.answer], key=lambda turn: turn.id)
    profile = session.extracted_profile or {}

    story = [
        _paragraph("RoleRAG Interview Report", title_style),
        _paragraph(
            f"Candidate-focused interview summary for the role {session.role}. "
            "This report captures profile extraction, turn-by-turn answers, evaluations, and the final interview summary.",
            body_style,
        ),
        Spacer(1, 8),
    ]

    metadata = Table(
        [
            [
                _paragraph("Candidate", label_style),
                _paragraph(profile.get("name") or session.candidate_name or "Unknown", body_style),
                _paragraph("Role", label_style),
                _paragraph(session.role, body_style),
            ],
            [
                _paragraph("Created", label_style),
                _paragraph(_format_datetime(session.created_at), body_style),
                _paragraph("Completed Turns", label_style),
                _paragraph(f"{len(completed_turns)} / {max_turns}", body_style),
            ],
            [
                _paragraph("Average Score", label_style),
                _paragraph(_average_score(completed_turns), body_style),
                _paragraph("Domains", label_style),
                _paragraph(_text(profile.get("domains")), body_style),
            ],
        ],
        colWidths=[1.15 * inch, 2.1 * inch, 1.2 * inch, 2.1 * inch],
    )
    metadata.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, -1), colors.HexColor("#f8fbff")),
                ("BOX", (0, 0), (-1, -1), 0.75, colors.HexColor("#d7dfeb")),
                ("INNERGRID", (0, 0), (-1, -1), 0.5, colors.HexColor("#d7dfeb")),
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 8),
                ("RIGHTPADDING", (0, 0), (-1, -1), 8),
                ("TOPPADDING", (0, 0), (-1, -1), 7),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 7),
            ]
        )
    )
    story.extend([metadata, Spacer(1, 14)])

    story.append(_paragraph("Extracted Candidate Profile", section_style))
    story.append(_labeled_paragraph("Skills", profile.get("skills"), body_style))
    story.append(_labeled_paragraph("Experience", profile.get("years_experience") or profile.get("seniority_signal"), body_style))
    story.append(_labeled_paragraph("Project Highlights", profile.get("projects"), body_style))
    story.append(_labeled_paragraph("Summary", profile.get("summary"), body_style))

    story.append(_paragraph("Final Interview Summary", section_style))
    story.append(_paragraph(session.summary or "Summary is not available yet for this session.", body_style))

    story.append(_paragraph("Turn-by-Turn Results", section_style))
    if not completed_turns:
        story.append(_paragraph("No completed interview turns were available when this report was generated.", body_style))
    else:
        for index, turn in enumerate(completed_turns, start=1):
            evaluation = turn.evaluation or {}
            sources = [item.get("metadata", {}).get("source") for item in (turn.retrieved_context or []) if item.get("metadata", {}).get("source")]

            story.append(_paragraph(f"Turn {index}", section_style))
            story.append(_labeled_paragraph("Question", turn.question, body_style))
            story.append(_labeled_paragraph("Answer", turn.answer, body_style))
            story.append(_labeled_paragraph("Score", evaluation.get("score"), body_style))
            story.append(_labeled_paragraph("Strengths", evaluation.get("strengths"), body_style))
            story.append(_labeled_paragraph("Gaps", evaluation.get("gaps"), body_style))
            story.append(_labeled_paragraph("Suggested Follow-up", evaluation.get("follow_up"), body_style))
            story.append(_labeled_paragraph("Grounding Sources", sources, small_style))
            story.append(Spacer(1, 6))

    story.append(Spacer(1, 8))
    story.append(_paragraph("Generated by RoleRAG Interviewer", small_style))

    doc.build(story)
    return buffer.getvalue()

File: netlify/functions/_lib/report.ts
import PDFDocument from 'pdfkit';
import type { InterviewSessionRow, InterviewTurnRow } from './types.js';

function text(value: unknown) {
  if (value == null) return '-';
  if (Array.isArray(value)) return value.filter(Boolean).join(', ') || '-';
  return String(value).trim() || '-';
}

function averageScore(turns: InterviewTurnRow[]) {
  const scores = turns
    .map((turn) => Number(turn.evaluation?.score))
    .filter((score) => Number.isFinite(score));
  if (!scores.length) return '-';
  return `${(scores.reduce((sum, score) => sum + score, 0) / scores.length).toFixed(1)} / 10`;
}

export function reportFilename(session: InterviewSessionRow) {
  const raw = session.candidate_name || session.extracted_profile?.name || `session-${session.id}`;
  const slug = raw.replace(/[^a-zA-Z0-9]+/g, '-').replace(/^-|-$/g, '').toLowerCase() || `session-${session.id}`;
  return `${slug}-interview-report.pdf`;
}

export async function buildSessionReport(session: InterviewSessionRow, turns: InterviewTurnRow[], maxTurns: number) {
  const completedTurns = turns.filter((turn) => turn.answer);
  const chunks: Buffer[] = [];
  const doc = new PDFDocument({ margin: 48, size: 'A4' });

  doc.on('data', (chunk: Buffer | Uint8Array) => chunks.push(Buffer.from(chunk)));
  const done = new Promise<Buffer>((resolve) => {
    doc.on('end', () => resolve(Buffer.concat(chunks)));
  });

  doc.fontSize(22).text('RoleRAG Interview Report', { underline: false });
  doc.moveDown(0.5);
  doc.fontSize(11).text(`Candidate-focused interview summary for the role ${session.role}.`);
  doc.moveDown();

  doc.fontSize(13).text('Session Metadata');
  doc.fontSize(10)
    .text(`Candidate: ${text(session.extracted_profile?.name || session.candidate_name)}`)
    .text(`Role: ${text(session.role)}`)
    .text(`Created: ${text(session.created_at)}`)
    .text(`Completed Turns: ${completedTurns.length} / ${maxTurns}`)
    .text(`Average Score: ${averageScore(completedTurns)}`);
  doc.moveDown();

  doc.fontSize(13).text('Extracted Candidate Profile');
  doc.fontSize(10)
    .text(`Skills: ${text(session.extracted_profile?.skills)}`)
    .text(`Experience: ${text(session.extracted_profile?.years_experience || session.extracted_profile?.seniority_signal)}`)
    .text(`Projects: ${text(session.extracted_profile?.projects)}`)
    .text(`Summary: ${text(session.extracted_profile?.summary)}`);
  doc.moveDown();

  doc.fontSize(13).text('Final Interview Summary');
  doc.fontSize(10).text(text(session.summary));
  doc.moveDown();

  doc.fontSize(13).text('Turn-by-Turn Results');
  for (const [index, turn] of completedTurns.entries()) {
    doc.moveDown(0.5);
    doc.fontSize(11).text(`Turn ${index + 1}`);
    doc.fontSize(10)
      .text(`Question: ${text(turn.question)}`)
      .text(`Answer: ${text(turn.answer)}`)
      .text(`Score: ${text(turn.evaluation?.score)}`)
      .text(`Strengths: ${text(turn.evaluation?.strengths)}`)
      .text(`Gaps: ${text(turn.evaluation?.gaps)}`)
      .text(`Suggested Follow-up: ${text(turn.evaluation?.follow_up)}`)
      .text(`Grounding Sources: ${text(turn.retrieved_context.map((item) => item.metadata.source))}`);
  }

  doc.end();
  return done;
}
